apm-agent-core/src/main/java/co/elastic/apm/agent/impl/async/SpanInScopeRunnableWrapper.java

```diff
@@ -45,6 +45,7 @@ public SpanInScopeRunnableWrapper(ElasticApmTracer tracer) {
     public SpanInScopeRunnableWrapper wrap(Runnable delegate, AbstractSpan<?> span) {
         this.delegate = delegate;
         this.span = span;
+        span.incrementReferences();
```

Comment:
In those cases the GC will take care of this transaction. We can't move it to run as run might be called after the span is ended and sent to the APM Server. If we don't increment the references when we wrap, indicating that we expect this reference to be used later, it might already be recycled when run starts.

Visualization:

```
█████
└─────────────█████
    ^         ^
    submit()  run()
    wrap()
```

Reply:
Not sure if an optimization for that (edge) case is worth the effort. I think we can add that later if we need. Maybe a better way would be to instrument the corresponding `Future`'s cancel method.